website/server/libs/webhook.js

```diff
@@ -1,31 +1,112 @@
-import { each } from 'lodash';
 import { post } from 'request';
 import { isURL } from 'validator';
 import logger from './logger';
+import common from '../../../common';
 
-let _sendWebhook = (url, body) => {
+function sendWebhook (url, body) {
   post({
     url,
     body,
     json: true,
-  }, (err) => logger.error(err));
-};
-
-let _isInvalidWebhook = (hook) => {
-  return !hook.enabled || !isURL(hook.url);
-};
-
-export function sendTaskWebhook (webhooks, data) {
-  each(webhooks, (hook) => {
-    if (_isInvalidWebhook(hook)) return;
-
-    let body = {
-      direction: data.task.direction,
-      task: data.task.details,
-      delta: data.task.delta,
-      user: data.user,
-    };
-
-    _sendWebhook(hook.url, body);
+  }, (err) => {
+    if (err) {
+      logger.error(err);
+    }
   });
 }
+
+function isValidWebhook (hook) {
+  return hook.enabled && isURL(hook.url);
+}
+
+export class WebhookSender {
+  constructor (options = {}) {
+    this.type = options.type;
+    this.transformData = options.transformData || WebhookSender.defaultTransformData;
+    this.webhookFilter = options.webhookFilter || WebhookSender.defaultWebhookFilter;
+  }
+
+  static defaultTransformData (data) {
+    return data;
+  }
+
+  static defaultWebhookFilter () {
+    return true;
+  }
+
+  send (webhooks, data) {
+    let hooks = Object.keys(webhooks)
+      .map(hook => webhooks[hook])
+      .filter(isValidWebhook)
+      // Until the migration to add the taskScored field to the webhooks is complete, we'll assume that any webhooks without a type are taskScored
+      // TODO: Change back to this when migration is complete
+      // .filter(hook => this.type === hook.type)
+      .filter((hook) => {
+        let type = hook.type || 'taskScored';
+        return this.type === type;
+      })
+      .filter(hook => this.webhookFilter(hook, data));
+
+    if (hooks.length < 1) {
+      return; // prevents running the body creation code if there are no webhooks to send
+    }
+
+    let body = this.transformData(data);
+
+    hooks.forEach((hook) => {
+      sendWebhook(hook.url, body);
+    });
+  }
+}
+
+export let taskScoredWebhook = new WebhookSender({
+  type: 'taskScored',
+  transformData (data) {
+    let { user, task, direction, delta } = data;
+
+    let extendedStats = Object.assign({
```

Comment:
I don't love either but from a pure perf point of view it's better to change the model's method to take a stats object